File: ruby/crash-lab/lib/crash_extension.rb
# frozen_string_literal: true

require "plushie"

# Crash widget extension -- deliberately panics on command.
#
# Renders a status label normally. When the "panic" command is sent,
# the Rust side calls panic!(), which the renderer catches via
# catch_unwind. The extension is marked "poisoned" and subsequent
# renders show a red error placeholder. Other widgets keep working.
#
# Events: none (the panic is one-way).
class CrashExtension
  include Plushie::Extension

  widget :crash_widget, kind: :native_widget

  rust_crate "native/crash_widget"
  rust_constructor "crash_widget::CrashExtension::new()"

  prop :label, :string, default: ""

  command :panic
end

File: elixir/crash-test/lib/crash_test/crash_extension.ex
defmodule CrashTest.CrashExtension do
  @moduledoc """
  Minimal native widget that can be deliberately panicked.

  Renders a green box with "Widget OK" text. Has a single `panic`
  command that calls `panic!()` in Rust's `handle_command`. After
  the panic, the renderer replaces this widget with a red placeholder
  via `catch_unwind` isolation.

  The widget stays broken for the rest of the session -
  demonstrating that the panic is per-widget-instance, not per-app.
  """

  use Plushie.Widget, :native_widget

  widget(:crash_widget)

  rust_crate("native/crash_widget")
  rust_constructor("crash_widget::CrashExtension::new()")

  prop(:label, :string, default: "Widget OK")

  command(:panic)
end
